File: parser/parser.go
package parser

import (
	"fmt"
	"gorb/ast"
	"gorb/lexer"
	"gorb/token"
	"strconv"
)

func New(l *lexer.Lexer) *Parser {
	p := &Parser{
		l:      l,
		errors: []string{},
	}

	p.unaryParseFns = make(map[token.TokenType]unaryParseFn)
	p.registerUnary(token.PAREN_OPEN, p.parseGroupedExpression)
	p.registerUnary(token.IF, p.parseIfExpression)
	p.registerUnary(token.NEGATE, p.parseUnaryExpression)
	p.registerUnary(token.MINUS, p.parseUnaryExpression)
	p.registerUnary(token.FUNCTION, p.parseFunctionLiteral)
	p.registerUnary(token.IDENTIFIER, p.parseIdentifier)
	p.registerUnary(token.TRUE, p.parseBoolean)
	p.registerUnary(token.FALSE, p.parseBoolean)
	p.registerUnary(token.INTEGER, p.parseIntegerLiteral)

	p.binaryParseFns = make(map[token.TokenType]binaryParseFn)
	p.registerBinary(token.PAREN_OPEN, p.parseCallExpression)
	p.registerBinary(token.PLUS, p.parseBinaryExpression)
	p.registerBinary(token.MINUS, p.parseBinaryExpression)
	p.registerBinary(token.SLASH, p.parseBinaryExpression)
	p.registerBinary(token.ASTERISK, p.parseBinaryExpression)
	p.registerBinary(token.EQUALS, p.parseBinaryExpression)
	p.registerBinary(token.NOT_EQUALS, p.parseBinaryExpression)
	p.registerBinary(token.LESS_THAN, p.parseBinaryExpression)
	p.registerBinary(token.GREATER_THAN, p.parseBinaryExpression)

	p.nextToken()
	p.nextToken()

	return p
}

type Parser struct {
	l      *lexer.Lexer
	errors []string

	curToken  token.Token
	peekToken token.Token

	unaryParseFns  map[token.TokenType]unaryParseFn
	binaryParseFns map[token.TokenType]binaryParseFn
}

type (
	unaryParseFn  func() ast.Expression
	binaryParseFn func(ast.Expression) ast.Expression
)

func (p *Parser) registerUnary(tt token.TokenType, fn unaryParseFn) {
	p.unaryParseFns[tt] = fn
}

func (p *Parser) registerBinary(tt token.TokenType, fn binaryParseFn) {
	p.binaryParseFns[tt] = fn
}

func (p *Parser) nextToken() {
	p.curToken = p.peekToken
	p.peekToken = p.l.NextToken()
}

func (p *Parser) ParseProgram() *ast.Program {
	program := &ast.Program{}
	program.Statements = []ast.Statement{}

	for !p.curTokenIs(token.EOF) {
		stmt := p.parseStatement()
		if stmt != nil {
			program.Statements = append(program.Statements, stmt)
		}
		p.nextToken()
	}

	return program
}

// statements

func (p *Parser) parseStatement() ast.Statement {
	switch p.curToken.Type {
	case token.RETURN:
		return p.parseReturnStatement()
	case token.DECLARATION:
		return p.parseDeclarationStatement()
	default:
		return p.parseExpressionStatement()
	}
}

func (p *Parser) parseReturnStatement() *ast.ReturnStatement {
	stmt := &ast.ReturnStatement{Token: p.curToken}

	p.nextToken()

	stmt.ReturnValue = p.parseExpression(LOWEST)

	for p.peekTokenIs(token.TERMINATOR) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseDeclarationStatement() *ast.DeclarationStatement {
	stmt := &ast.DeclarationStatement{Token: p.curToken}

	if !p.expectPeek(token.IDENTIFIER) {
		return nil
	}

	stmt.Name = &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}

	if !p.expectPeek(token.ASSIGN) {
		return nil
	}

	p.nextToken()

	stmt.Value = p.parseExpression(LOWEST)

	for p.peekTokenIs(token.TERMINATOR) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseExpressionStatement() *ast.ExpressionStatement {
	stmt := &ast.ExpressionStatement{Token: p.curToken}

	stmt.Expression = p.parseExpression(LOWEST)

	for p.peekTokenIs(token.TERMINATOR) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseBlockStatement() *ast.BlockStatement {
	block := &ast.BlockStatement{Token: p.curToken}
	block.Statements = []ast.Statement{}

	p.nextToken()

	for !p.curTokenIs(token.BRACE_CLOSE) && !p.curTokenIs(token.EOF) {
		stmt := p.parseStatement()
		if stmt != nil {
			block.Statements = append(block.Statements, stmt)
		}
		p.nextToken()
	}

	return block
}

// expressions

func (p *Parser) parseGroupedExpression() ast.Expression {
	p.nextToken()

	exp := p.parseExpression(LOWEST)

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	return exp
}

func (p *Parser) parseCallExpression(function ast.Expression) ast.Expression {
	exp := &ast.CallExpression{Token: p.curToken, Function: function}
	exp.Arguments = p.parseCallArguments()

	return exp
}

func (p *Parser) parseCallArguments() []ast.Expression {
	args := []ast.Expression{}

	if p.peekTokenIs(token.PAREN_CLOSE) {
		p.nextToken()
		return args
	}

	p.nextToken()

	args = append(args, p.parseExpression(LOWEST))

	for p.peekTokenIs(token.COMMA) {
		p.nextToken()
		p.nextToken()
		args = append(args, p.parseExpression(LOWEST))
	}

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	return args
}

func (p *Parser) parseIfExpression() ast.Expression {
	exp := &ast.IfExpression{Token: p.curToken}

	if !p.expectPeek(token.PAREN_OPEN) {
		return nil
	}

	p.nextToken()
	exp.Condition = p.parseExpression(LOWEST)

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	if !p.expectPeek(token.BRACE_OPEN) {
		return nil
	}

	exp.Consequence = p.parseBlockStatement()

	if p.peekTokenIs(token.ELSE) {
		p.nextToken()

		if !p.expectPeek(token.BRACE_OPEN) {
			return nil
		}

		exp.Alternative = p.parseBlockStatement()
	}

	return exp
}

func (p *Parser) parseExpression(precedence int) ast.Expression {
	parseUnary := p.unaryParseFns[p.curToken.Type]
	if parseUnary == nil {
		p.noUnaryParseFnError(p.curToken.Type)
		return nil
	}
	leftExp := parseUnary()

	for !p.peekTokenIs(token.TERMINATOR) && precedence < p.peekPrecedence() {
		parseBinary := p.binaryParseFns[p.peekToken.Type]
		if parseBinary == nil {
			return leftExp
		}

		p.nextToken()

		leftExp = parseBinary(leftExp)
	}

	return leftExp
}

func (p *Parser) parseUnaryExpression() ast.Expression {
	exp := &ast.UnaryExpression{
		Token:    p.curToken,
		Operator: p.curToken.Type,
	}

	p.nextToken()

	exp.Right = p.parseExpression(UNARY)

	return exp
}

func (p *Parser) parseBinaryExpression(left ast.Expression) ast.Expression {
	exp := &ast.BinaryExpression{
		Token:    p.curToken,
		Operator: p.curToken.Type,
		Left:     left,
	}

	precedence := p.curPrecedence()
	p.nextToken()
	exp.Right = p.parseExpression(precedence)

	return exp
}

// literals

func (p *Parser) parseFunctionLiteral() ast.Expression {
	lit := &ast.FunctionLiteral{Token: p.curToken}

	if !p.expectPeek(token.PAREN_OPEN) {
		return nil
	}

	lit.Parameters = p.parseFunctionParameters()

	if !p.expectPeek(token.BRACE_OPEN) {
		return nil
	}

	lit.Body = p.parseBlockStatement()

	return lit
}

func (p *Parser) parseFunctionParameters() []*ast.Identifier {
	identifiers := []*ast.Identifier{}

	if p.peekTokenIs(token.PAREN_CLOSE) {
		p.nextToken()
		return identifiers
	}

	p.nextToken()

	ident := &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
	identifiers = append(identifiers, ident)

	for p.peekTokenIs(token.COMMA) {
		p.nextToken()
		p.nextToken()
		ident := &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
		identifiers = append(identifiers, ident)
	}

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	return identifiers
}

func (p *Parser) parseIdentifier() ast.Expression {
	return &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
}

func (p *Parser) parseBoolean() ast.Expression {
	return &ast.BooleanLiteral{Token: p.curToken, Value: p.curTokenIs(token.TRUE)}
}

func (p *Parser) parseIntegerLiteral() ast.Expression {
	lit := &ast.IntegerLiteral{Token: p.curToken}

	val, err := strconv.ParseInt(p.curToken.Literal, 0, 64)

	if err != nil {
		msg := fmt.Sprintf(
			"%v could not parse %q as integer",
			p.curToken.Pos,
			p.curToken.Literal,
		)
		p.errors = append(p.errors, msg)
		return nil
	}

	lit.Value = val

	return lit
}

// helpers

func (p *Parser) curTokenIs(tt token.TokenType) bool {
	return p.curToken.Type == tt
}

func (p *Parser) peekTokenIs(tt token.TokenType) bool {
	return p.peekToken.Type == tt
}

func (p *Parser) expectPeek(tt token.TokenType) bool {
	if p.peekTokenIs(tt) {
		p.nextToken()
		return true
	} else {
		p.peekError(tt)
		return false
	}
}

const (
	_ int = iota
	LOWEST
	EQUALITY
	COMPARISON
	SUM
	PRODUCT
	UNARY
	CALL
)

var precedences = map[token.TokenType]int{
	token.EQUALS:       EQUALITY,
	token.NOT_EQUALS:   EQUALITY,
	token.LESS_THAN:    COMPARISON,
	token.GREATER_THAN: COMPARISON,
	token.PLUS:          SUM,
	token.MINUS:     SUM,
	token.ASTERISK:     PRODUCT,
	token.SLASH:       PRODUCT,
	token.PAREN_OPEN:   CALL,
}

func (p *Parser) peekPrecedence() int {
	if p, ok := precedences[p.peekToken.Type]; ok {
		return p
	}

	return LOWEST
}

func (p *Parser) curPrecedence() int {
	if p, ok := precedences[p.curToken.Type]; ok {
		return p
	}

	return LOWEST
}

// errors

func (p *Parser) Errors() []string {
	return p.errors
}

func (p *Parser) peekError(t token.TokenType) {
	msg := fmt.Sprintf(
		"%v expected %s, got %s",
		p.curToken.Pos,
		t,
		p.peekToken.Type,
	)
	p.errors = append(p.errors, msg)
}

func (p *Parser) noUnaryParseFnError(t token.TokenType) {
	msg := fmt.Sprintf(
		"%v no unary parse function for %s found",
		p.curToken.Pos,
		t,
	)
	p.errors = append(p.errors, msg)
}

File: src/Parser.zig
const std = @import("std");
const Lexer = @import("./Lexer.zig");
const Token = @import("./Token.zig");
const ast = @import("./ast.zig");
const errors = @import("./errors.zig");
const object = @import("./object.zig");

lexer: Lexer,
errors: std.ArrayList([]const u8),

cur_token: Token = undefined,
peek_token: Token = undefined,

allocator: std.mem.Allocator,

const Self = @This();

pub fn init(
    allocator: std.mem.Allocator,
    l: Lexer,
) Self {
    var parser = Self{
        .lexer = l,
        .errors = std.ArrayList([]const u8).init(allocator),
        .allocator = allocator,
    };

    parser.nextToken();
    parser.nextToken();

    return parser;
}

pub fn deinit(self: *Self) void {
    for (self.errors.items) |err| self.allocator.free(err);
    self.errors.deinit();
}

pub fn parseProgram(self: *Self) !ast.Node {
    const program = try ast.Block.init(self.allocator);
    var statements = std.ArrayList(ast.Node).init(self.allocator);
    errdefer {
        for (statements.items) |stmt| stmt.deinit(self.allocator);
        statements.deinit();
        self.peek_token.deinit(self.allocator);
        self.allocator.destroy(program);
    }

    while (!self.curTokenIs(.eof)) {
        const stmt = try self.parseStatement();
        try statements.append(stmt);
        self.nextToken();
    }

    program.statements = try statements.toOwnedSlice();

    return .{ .block = program };
}

// statements

fn parseStatement(self: *Self) !ast.Node {
    return switch (self.cur_token.type) {
        .kw_return => try self.parseReturnStatement(),
        .kw_declaration => try self.parseDeclarationStatement(),
        else => try self.parseExpressionStatement(),
    };
}

fn parseReturnStatement(self: *Self) !ast.Node {
    const stmt = try ast.Return.init(self.allocator);
    errdefer self.allocator.destroy(stmt);
    stmt.tok = self.cur_token;

    self.nextToken();

    stmt.return_value = try self.parseExpression(.lowest);

    while (self.peekTokenIs(.semicolon)) {
        self.nextToken();
    }

    return .{ .@"return" = stmt };
}

fn parseDeclarationStatement(self: *Self) !ast.Node {
    const stmt = try ast.Declaration.init(self.allocator);
    errdefer self.allocator.destroy(stmt);
    stmt.tok = self.cur_token;

    try self.expectPeek(.identifier);

    const identifier = try ast.Identifier.init(self.allocator);
    errdefer self.allocator.destroy(identifier);
    identifier.tok = self.cur_token;
    identifier.value = self.cur_token.literal;

    stmt.name = identifier;

    try self.expectPeek(.assign);

    self.nextToken();

    stmt.value = try self.parseExpression(.lowest);

    while (self.peekTokenIs(.semicolon)) {
        self.nextToken();
    }

    return .{ .declaration = stmt };
}

fn parseExpressionStatement(self: *Self) !ast.Node {
    const node = self.parseExpression(.lowest);

    while (self.peekTokenIs(.semicolon)) {
        self.nextToken();
    }

    return node;
}

fn parseBlockStatement(self: *Self) !ast.Node {
    const block = try ast.Block.init(self.allocator);
    errdefer self.allocator.destroy(block);
    var statements = std.ArrayList(ast.Node).init(self.allocator);
    errdefer {
        for (statements.items) |stmt| stmt.deinit(self.allocator);
        statements.deinit();
    }

    self.nextToken();

    while (!self.curTokenIs(.brace_close)) {
        const stmt = try self.parseStatement();
        try statements.append(stmt);
        self.nextToken();
    }

    block.statements = try statements.toOwnedSlice();

    return .{ .block = block };
}

// expressions

fn parseExpression(
    self: *Self,
    precedence: Precedence,
) Error!ast.Node {
    var expr: ast.Node = switch (self.cur_token.type) {
        .paren_open => try self.parseGroupedExpression(),
        .kw_if => try self.parseIfExpression(),
        .bang => try self.parseUnaryExpression(),
        .minus => try self.parseUnaryExpression(),
        .kw_function => try self.parseFunctionLiteral(),
        .identifier => try self.parseIdentifier(),
        .kw_true => try self.parseBooleanLiteral(),
        .kw_false => try self.parseBooleanLiteral(),
        .integer => try self.parseIntegerLiteral(),
        .string => try self.parseStringLiteral(),
        .bracket_open => try self.parseArrayLiteral(),
        .brace_open => try self.parseHashLiteral(),
        else => {
            try self.noUnaryParseFnError(self.cur_token);
            return ParserError.NoUnaryParseFn;
        },
    };
    errdefer expr.deinit(self.allocator);

    while (!self.peekTokenIs(.semicolon) and @intFromEnum(precedence) < @intFromEnum(self.peekPrecedence())) {
        const tok_type = self.peek_token.type;
        self.nextToken();
        expr = switch (tok_type) {
            .bracket_open => try self.parseIndexExpression(expr),
            .paren_open => try self.parseCallExpression(expr),
            .plus => try self.parseBinaryExpression(expr),
            .minus => try self.parseBinaryExpression(expr),
            .slash => try self.parseBinaryExpression(expr),
            .asterisk => try self.parseBinaryExpression(expr),
            .equals => try self.parseBinaryExpression(expr),
            .not_equals => try self.parseBinaryExpression(expr),
            .less_than => try self.parseBinaryExpression(expr),
            .greater_than => try self.parseBinaryExpression(expr),
            else => return expr,
        };
    }

    return expr;
}

fn parseGroupedExpression(self: *Self) Error!ast.Node {
    self.nextToken();

    const node = try self.parseExpression(.lowest);
    errdefer node.deinit(self.allocator);

    try self.expectPeek(.paren_close);

    return node;
}

fn parseIndexExpression(self: *Self, left: ast.Node) !ast.Node {
    const expr = try ast.Index.init(self.allocator);
    errdefer self.allocator.destroy(expr);
    expr.tok = self.cur_token;
    expr.left = left;

    self.nextToken();
    expr.index = try self.parseExpression(.lowest);
    errdefer expr.index.deinit(self.allocator);

    try self.expectPeek(.bracket_close);

    return .{ .index = expr };
}

fn parseCallExpression(
    self: *Self,
    function: ast.Node,
) !ast.Node {
    const expr = try ast.Call.init(self.allocator);
    errdefer self.allocator.destroy(expr);
    expr.tok = self.cur_token;
    expr.function = function;
    expr.arguments = try self.parseExpressionList(.paren_close);

    return .{ .call = expr };
}

fn parseIfExpression(self: *Self) !ast.Node {
    const expr = try ast.If.init(self.allocator);
    errdefer self.allocator.destroy(expr);
    expr.tok = self.cur_token;

    try self.expectPeek(.paren_open);

    self.nextToken();
    expr.condition = try self.parseExpression(.lowest);
    errdefer expr.condition.deinit(self.allocator);

    try self.expectPeek(.paren_close);

    try self.expectPeek(.brace_open);

    expr.consequence = (try self.parseBlockStatement()).block;
    errdefer expr.consequence.deinit(self.allocator);

    if (self.peekTokenIs(.kw_else)) {
        self.nextToken();

        try self.expectPeek(.brace_open);

        expr.alternative = (try self.parseBlockStatement()).block;
    } else {
        expr.alternative = null;
    }

    return .{ .@"if" = expr };
}

fn parseUnaryExpression(self: *Self) !ast.Node {
    const expr = try ast.UnaryOperation.init(self.allocator);
    errdefer self.allocator.destroy(expr);
    expr.tok = self.cur_token;
    expr.operator = self.cur_token;

    self.nextToken();

    expr.right = try self.parseExpression(.unary);

    return .{ .unary_operation = expr };
}

fn parseBinaryExpression(
    self: *Self,
    left: ast.Node,
) !ast.Node {
    const expr = try ast.BinaryOperation.init(self.allocator);
    errdefer self.allocator.destroy(expr);
    expr.tok = self.cur_token;
    expr.operator = self.cur_token;
    expr.left = left;

    const precedence = self.curPrecedence();
    self.nextToken();
    expr.right = try self.parseExpression(precedence);

    return .{ .binary_operation = expr };
}

fn parseExpressionList(
    self: *Self,
    end: Token.TokenType,
) ![]ast.Node {
    var list = std.ArrayList(ast.Node).init(self.allocator);
    errdefer {
        for (list.items) |expr| expr.deinit(self.allocator);
        list.deinit();
    }

    if (self.peekTokenIs(end)) {
        self.nextToken();
        return try list.toOwnedSlice();
    }

    self.nextToken();
    try list.append(try self.parseExpression(.lowest));

    while (self.peekTokenIs(.comma)) {
        self.nextToken();
        self.nextToken();
        try list.append(try self.parseExpression(.lowest));
    }

    try self.expectPeek(end);

    return try list.toOwnedSlice();
}

// literals

fn parseIdentifier(self: *Self) !ast.Node {
    const identifier = try ast.Identifier.init(self.allocator);
    errdefer self.allocator.destroy(identifier);
    identifier.tok = self.cur_token;
    identifier.value = self.cur_token.literal;
    return .{ .identifier = identifier };
}

fn parseBooleanLiteral(self: *Self) !ast.Node {
    const boolean = try ast.BooleanLiteral.init(self.allocator);
    errdefer self.allocator.destroy(boolean);
    boolean.tok = self.cur_token;
    boolean.value = self.curTokenIs(Token.TokenType.kw_true);
    return .{ .boolean_literal = boolean };
}

fn parseIntegerLiteral(self: *Self) !ast.Node {
    const integer = try ast.IntegerLiteral.init(self.allocator);
    errdefer self.allocator.destroy(integer);
    integer.tok = self.cur_token;
    integer.value = try std.fmt.parseInt(
        object.Integer,
        self.cur_token.literal,
        10,
    );
    return .{ .integer_literal = integer };
}

fn parseStringLiteral(self: *Self) !ast.Node {
    const string = try ast.StringLiteral.init(self.allocator);
    errdefer self.allocator.destroy(string);
    string.tok = self.cur_token;
    string.value = self.cur_token.literal;
    return .{ .string_literal = string };
}

fn parseArrayLiteral(self: *Self) !ast.Node {
    const array = try ast.ArrayLiteral.init(self.allocator);
    errdefer self.allocator.destroy(array);
    array.tok = self.cur_token;
    array.elements = try self.parseExpressionList(
        .bracket_close,
    );
    return .{ .array_literal = array };
}

fn parseHashLiteral(self: *Self) !ast.Node {
    const hash = try ast.HashLiteral.init(self.allocator);
    errdefer self.allocator.destroy(hash);
    hash.tok = self.cur_token;
    var pairs = std.AutoHashMap(ast.Node, ast.Node).init(self.allocator);
    errdefer {
        var iterator = pairs.iterator();
        while (iterator.next()) |value| {
            value.key_ptr.deinit(self.allocator);
            value.value_ptr.deinit(self.allocator);
        }
        pairs.deinit();
    }

    while (!self.peekTokenIs(.brace_close)) {
        self.nextToken();
        const key = try self.parseExpression(.lowest);
        errdefer key.deinit(self.allocator);

        try self.expectPeek(.colon);

        self.nextToken();
        const value = try self.parseExpression(.lowest);
        errdefer value.deinit(self.allocator);

        try pairs.put(key, value);

        if (!self.peekTokenIs(.brace_close)) {
            try self.expectPeek(.comma);
        }
    }

    try self.expectPeek(.brace_close);

    hash.pairs = pairs;
    return .{ .hash_literal = hash };
}

fn parseFunctionLiteral(self: *Self) !ast.Node {
    const function = try ast.FunctionLiteral.init(self.allocator);
    errdefer self.allocator.destroy(function);
    function.tok = self.cur_token;

    try self.expectPeek(.paren_open);

    function.parameters = try self.parseFunctionParameters();
    errdefer {
        for (function.parameters) |param| param.deinit(self.allocator);
        self.allocator.free(function.parameters);
    }

    try self.expectPeek(.brace_open);

    function.body = (try self.parseBlockStatement()).block;

    return .{ .function_literal = function };
}

fn parseFunctionParameters(self: *Self) ![]*ast.Identifier {
    var identifiers = std.ArrayList(*ast.Identifier).init(self.allocator);
    errdefer {
        for (identifiers.items) |ident| ident.deinit(self.allocator);
        identifiers.deinit();
    }

    if (self.peekTokenIs(.paren_close)) {
        self.nextToken();
        return try identifiers.toOwnedSlice();
    }

    self.nextToken();

    const identifier = try ast.Identifier.init(self.allocator);
    identifier.tok = self.cur_token;
    identifier.value = self.cur_token.literal;
    try identifiers.append(identifier);

    while (self.peekTokenIs(.comma)) {
        self.nextToken();
        self.nextToken();

        const ident = try ast.Identifier.init(self.allocator);
        ident.tok = self.cur_token;
        ident.value = self.cur_token.literal;
        try identifiers.append(ident);
    }

    try self.expectPeek(.paren_close);

    return try identifiers.toOwnedSlice();
}

// helpers

fn nextToken(self: *Self) void {
    self.cur_token = self.peek_token;
    self.peek_token = self.lexer.nextToken() catch @panic("parser nextToken OOM");
}

fn curTokenIs(self: *Self, tok_type: Token.TokenType) bool {
    return self.cur_token.type == tok_type;
}

fn peekTokenIs(self: *Self, tok_type: Token.TokenType) bool {
    return self.peek_token.type == tok_type;
}

fn expectPeek(self: *Self, tok_type: Token.TokenType) !void {
    if (self.peekTokenIs(tok_type)) {
        self.nextToken();
    } else {
        try self.expectPeekError(tok_type);
        return ParserError.UnexpectedPeekToken;
    }
}

const Precedence = enum(u32) {
    lowest,
    equality,
    comparison,
    sum,
    product,
    unary,
    call,
    index,
};

fn getPrecedence(tok_type: Token.TokenType) Precedence {
    return switch (tok_type) {
        .equals => .equality,
        .not_equals => .equality,
        .less_than => .comparison,
        .greater_than => .comparison,
        .plus => .sum,
        .minus => .sum,
        .asterisk => .product,
        .slash => .product,
        .paren_open => .call,
        .bracket_open => .index,
        else => .lowest,
    };
}

fn peekPrecedence(self: *Self) Precedence {
    return getPrecedence(self.peek_token.type);
}

fn curPrecedence(self: *Self) Precedence {
    return getPrecedence(self.cur_token.type);
}

// errors

const Error = error{
    OutOfMemory,
    Overflow,
    InvalidCharacter,
} || ParserError;

pub const ParserError = error{
    UnexpectedPeekToken,
    NoUnaryParseFn,
};

fn newParserError(
    self: *Self,
    msg: []const u8,
    tok: Token,
) !void {
    defer self.allocator.free(msg);
    const message = errors.formatError(self.allocator, msg, tok);
    try self.errors.append(message);
}

fn expectPeekError(
    self: *Self,
    tok_type: Token.TokenType,
) !void {
    const tok_text = tok_type.fmt(self.allocator);
    const peek_tok_text = self.peek_token.fmt(self.allocator);
    defer self.allocator.free(tok_text);
    defer self.allocator.free(peek_tok_text);
    const msg = try std.fmt.allocPrint(
        self.allocator,
        "expected {s}, got {s}",
        .{
            tok_text,
            peek_tok_text,
        },
    );
    try self.newParserError(msg, self.peek_token);
}

fn noUnaryParseFnError(self: *Self, tok: Token) !void {
    const tok_text = tok.fmt(self.allocator);
    defer self.allocator.free(tok_text);
    const msg = try std.fmt.allocPrint(
        self.allocator,
        "no unary parse function for {s} found",
        .{
            tok_text,
        },
    );
    try self.newParserError(msg, tok);
}

test {
    _ = @import("./tests/parser_test.zig");
}
